ledger/participant-integration-api/src/test/lib/scala/platform/indexer/ha/TestDBLockStorageBackend.scala

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.indexer.ha
+
+import java.sql.{
+  Blob,
+  CallableStatement,
+  Clob,
+  Connection,
+  DatabaseMetaData,
+  NClob,
+  PreparedStatement,
+  SQLWarning,
+  SQLXML,
+  Savepoint,
+  Statement,
+  Struct,
+}
+import java.util.Properties
+import java.util.concurrent.Executor
+import java.{sql, util}
+
+import com.daml.platform.store.backend.DBLockStorageBackend
+
+class TestDBLockStorageBackend extends DBLockStorageBackend {
+
+  private var locks: Map[DBLockStorageBackend.Lock, Set[Connection]] = Map.empty
+
+  override def tryAcquire(
+      lockId: DBLockStorageBackend.LockId,
+      lockMode: DBLockStorageBackend.LockMode,
+  )(connection: Connection): Option[DBLockStorageBackend.Lock] = synchronized {
+    if (connection.isClosed) throw new Exception("trying to acquire on a closed connection")
+    if (!lockId.isInstanceOf[TestLockId]) throw new Exception("foreign lockId")
+    removeClosedConnectionRelatedLocks()
+    val lock = DBLockStorageBackend.Lock(lockId, lockMode)
+    def doLock(): Option[DBLockStorageBackend.Lock] = {
+      locks = locks + (lock -> (locks.getOrElse(lock, Set.empty) + connection))
+      Some(lock)
+    }
+    lockMode match {
+      case DBLockStorageBackend.LockMode.Exclusive =>
+        (
+          locks.get(lock),
+          locks.get(lock.copy(lockMode = DBLockStorageBackend.LockMode.Shared)),
+        ) match {
+          case (None, None) => doLock()
+          case (Some(connections), None) if connections == Set(connection) => doLock()
+          case _ => None // if any shared lock held, we cannot lock exclusively
+        }
+      case DBLockStorageBackend.LockMode.Shared =>
+        (
+          locks.get(lock),
+          locks.get(lock.copy(lockMode = DBLockStorageBackend.LockMode.Exclusive)),
+        ) match {
+          case (_, None) => doLock()
+          case _ => None // if any exclusive lock held, we cannot lock shared
+        }
+    }
+  }
+
+  override def release(lock: DBLockStorageBackend.Lock)(connection: Connection): Boolean =
+    synchronized {
+      if (connection.isClosed) throw new Exception("trying to release on a closed connection")
+      if (!lock.lockId.isInstanceOf[TestLockId]) throw new Exception("foreign lockId")
+      removeClosedConnectionRelatedLocks()
+      locks.get(lock) match {
+        case None => false
+        case Some(connections) if connections contains connection =>
+          if (connections.size == 1) locks = locks - lock
+          else locks = locks + (lock -> (connections - connection))
+          true
+        case _ => false
+      }
+    }
+
+  private def removeClosedConnectionRelatedLocks(): Unit =
+    locks = locks.collect {
+      case (lock, conns) if conns.exists(!_.isClosed) => (lock, conns.filter(!_.isClosed))
```

Comment:
the higher order function here is a collect, for that collect it is important to filter out locks which have no non-closed connections, since having an entry in the map with empty set is not equivalent to not having that entry

Reply:
ah, my Haskell background led me astray. I didn't realize that there is an implicit discarding happening here. Makes sense now.